nx/lib/nx.ex

```diff
@@ -7046,6 +7046,152 @@ defmodule Nx do
+      [12.0, -51.0, 4.0],
+      [6.0, 167.0, -68.0],
+      [-4.0, 24.0, -41.0]
+    ])
+
+    iex(1)> [q, r] = Nx.qr(Nx.tensor([[1, -1, 4], [1, 4, -2], [1, 4, 2], [1, -1, 0]]), mode: :reduced) |> Tuple.to_list() |> Enum.map(fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    [
+      Nx.tensor([
+        [-0.5774, 0.8165, 0.0],
+        [-0.5774, -0.4082, 0.7071],
+        [-0.5774, -0.4082, -0.7071],
+        [0.0, 0.0, 0.0]
+      ]),
+      Nx.tensor([
+        [-1.7321, -4.0415, -2.3094],
+        [0.0, -4.0825, 3.266],
+        [0.0, 0.0, -2.8284]
+      ])
+    ]
+    iex(2)> Nx.dot(q, r) |> Nx.map(&Float.round(&1, 0))
+    Nx.tensor([
+      [1.0, -1.0, 4.0],
+      [1.0, 4.0, -2.0],
+      [1.0, 4.0, 2.0],
+      [0.0, -0.0, 0.0]
+    ])
+
+    iex(1)> [q, r] = Nx.qr(Nx.tensor([[1, -1, 4], [1, 4, -2], [1, 4, 2], [1, -1, 0]]), mode: :complete) |> Tuple.to_list() |> Enum.map(fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    [
+      Nx.tensor([
+        [-0.5, 0.5, -0.5, -0.5],
+        [-0.5, -0.5, 0.5, -0.5],
+        [-0.5, -0.5, -0.5, 0.5],
+        [-0.5, 0.5, 0.5, 0.5]
+      ]),
+      Nx.tensor([
+        [-2.0, -3.0, -2.0],
+        [0.0, -5.0, 2.0],
+        [-0.0, -0.0, -4.0],
+        [0.0, 0.0, 0.0]
+      ])
+    ]
+    iex(2)> Nx.dot(q, r) |> Nx.map(&Float.round(&1, 0))
+    Nx.tensor([
+      [1.0, -1.0, 4.0],
+      [1.0, 4.0, -2.0],
+      [1.0, 4.0, 2.0],
+      [1.0, -1.0, 0.0]
+    ])
+
+    iex(1)> {_q, _r, reflectors} = Nx.qr(Nx.tensor([[1, -1, 4], [1, 4, -2], [1, 4, 2], [1, -1, 0]]), mode: :raw)
+    iex(2)> Enum.map(reflectors, fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    [
+      Nx.tensor([
+        [1.0, 0.0, 0.0, 0.0],
+        [0.0, 1.0, 0.0, 0.0],
+        [0.0, 0.0, -0.6, 0.8],
+        [0.0, 0.0, 0.8, 0.6]
+      ]),
+      Nx.tensor([
+        [1.0, 0.0, 0.0, 0.0],
+        [0.0, -0.6667, -0.6667, 0.3333],
+        [0.0, -0.6667, 0.7333, 0.1333],
+        [0.0, 0.3333, 0.1333, 0.9333]
+      ]),
+      Nx.tensor([
+        [-0.5, -0.5, -0.5, -0.5],
+        [-0.5, 0.8333, -0.1667, -0.1667],
+        [-0.5, -0.1667, 0.8333, -0.1667],
+        [-0.5, -0.1667, -0.1667, 0.8333]
+      ])
+    ]
+
+  ## Error cases
+
+    iex> Nx.qr(Nx.tensor([[1, 1, 1, 1], [-1, 4, 4, -1], [4, -2, 2, 0]])) |> Tuple.to_list() |> Enum.map(fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    ** (ArgumentError) tensor must have at least as many rows than columns, got shape: {3, 4}
+  """
+  @doc type: :linalg
+  def qr(tensor, opts \\ []) do
+    %T{type: type, shape: shape} = tensor = tensor!(tensor)
+
+    assert_keys!(opts, [:mode])
+
+    opts = Keyword.merge([mode: :reduced], opts)
+
+    mode = opts[:mode]
+    valid_modes = [:reduced, :complete, :r, :raw]
+
+    unless mode in valid_modes do
+      raise ArgumentError,
+            "invalid :mode received. Expected one of #{valid_modes}, received: #{mode}"
+    end
+
+    output_type = Nx.Type.to_floating(type)
+
+    {q_shape, {k, n}} = Nx.Shape.qr(shape, opts)
```

Comment:
Nice, note you should also include a name rule here. Because this is a factorization of `A` into the product `QR`, the names should inherit from `A` such that the names of `Q . R` are equal to the names in `A`. I think it would be something like:

```
a_names = [n1, n2]
q_names = [n1, nil]
r_names = [nil, n2]
```

And perhaps we can provide an option for naming those unnamed dimensions? The other option is to just set names of the new tensor to nil

Reply:
Sounds like a good idea! I think that keeping the names when they are given is good. Should we require names for the unnamed dimensions, or maybe provide default ones when input names are given?